router/router.go

```diff
@@ -0,0 +1,153 @@
+package router
+
+import (
+	"github.com/ONSdigital/dp-frontend-router/middleware/allRoutes"
+	"github.com/ONSdigital/dp-frontend-router/middleware/redirects"
+	"github.com/ONSdigital/dp-frontend-router/middleware/serverError"
+	dprequest "github.com/ONSdigital/dp-net/request"
+	"github.com/ONSdigital/log.go/log"
+	"github.com/gorilla/mux"
+	"github.com/gorilla/pat"
+	"github.com/justinas/alice"
+	"net/http"
+	"path/filepath"
+	"strings"
+)
+
+//go:generate moq -out routertest/handler.go -pkg routertest . Handler
+type Handler http.Handler
+
+type Config struct {
+	HealthCheckHandler   func(w http.ResponseWriter, req *http.Request)
+	AnalyticsHandler     http.Handler
+	DownloadHandler      http.Handler
+	DatasetHandler       http.Handler
+	CookieHandler        http.Handler
+	FilterHandler        http.Handler
+	FeedbackHandler      http.Handler
+	ContentTypeByteLimit int
+	ZebedeeClient        allRoutes.ZebedeeClient
+	GeographyEnabled     bool
+	GeographyHandler     http.Handler
+	SearchRoutesEnabled  bool
+	SearchHandler        http.Handler
+	HomepageHandler      http.Handler
+	BabbageHandler       http.Handler
+}
+
+func New(cfg Config) http.Handler {
+
+	router := pat.New()
+
+	middleware := []alice.Constructor{
+		dprequest.HandlerRequestID(16),
+		log.Middleware,
+		securityHandler,
+		healthcheckHandler(cfg.HealthCheckHandler),
+		serverError.Handler,
+		redirects.Handler,
+	}
+
+	alice := alice.New(middleware...).Then(router)
+
+	router.Handle("/redir/{data:.*}", cfg.AnalyticsHandler)
+	router.Handle("/download/{uri:.*}", cfg.DownloadHandler)
+	router.Handle("/cookies{uri:.*}", cfg.CookieHandler)
+	router.Handle("/datasets/{uri:.*}", cfg.DatasetHandler)
+	router.Handle("/filters/{uri:.*}", cfg.FilterHandler)
+	router.Handle("/filter-outputs/{uri:.*}", cfg.FilterHandler)
+	router.Handle("/feedback{uri:.*}", cfg.FeedbackHandler)
+
+	if cfg.GeographyEnabled {
+		router.Handle("/geography{uri:.*}", cfg.GeographyHandler)
+	}
+
+	if cfg.SearchRoutesEnabled {
+		router.Handle("/search", cfg.SearchHandler)
+	}
+
+	router.Handle("/", cfg.HomepageHandler)
```

Comment:
Good point - I have moved it to the top. It would be good to gather some metrics on which endpoints are most popular so we can tweak the ordering and save some CPU if needed.